### PR TITLE
[RFC] Interruptable scope_graph generation

### PR DESCRIPTION
**Pull request by rsdy**
Reduce the parsing time to 0.1s. For some reason the previous 1s limit causes weird runaway loops parsing highly complex files. We might want to escalate this to upstream with a test suite.

# Verifying results

[983d36c](https://github.com/BloopAI/bloop/pull/27/commits/983d36c3caf965396bbcfb4a7077988c2dac9604) reduces the parsing timeout to 0.1s on paper. This somehow avoids choking the indexing process, and extreme outliers for the entire worker function. Ultimately it cuts total processing from 7m40s to 3m20s for the testset.

### CODE REVIEW

**Comment by rsdy** [comment]
Profiling the bottlenecked parts that are constrained to a single thread, using `perf`, shows the following:

```
47.86%  bleep            bleep              [.] ts_tree_cursor_parent_node        
  41.14%  bleep            bleep              [.] ts_tree_cursor_current_status     
  10.60%  bleep            bleep              [.] ts_language_symbol_metadata       
   0.06%  bleep            bleep              [.] ts_query_cursor__advance    
```

Trying to repro on macOS with additional debug info, the call stack is not proving super helpful:
Screenshot 2022-12-14 at 21 44 50

There seems to be a tight loop somewhere we can't interrupt.

**Comment by calyptobai** [comment]
Interesting approach. I'm curious if we can fix the performance issue in an alternate way with the use of a new data structure? If there's no obvious fix, and if this does show better performance in tests, it seems like a decent stopgap.

**Comment by rsdy** [comment]
[983d36c](https://github.com/BloopAI/bloop/pull/27/commits/983d36c3caf965396bbcfb4a7077988c2dac9604) reduces the parsing timeout to 0.1s on paper. This somehow avoids choking the indexing process, and extreme outliers for the entire `worker` function.

Unclear if we need the rest of the hacks for safety.

File: server/bleep/src/intelligence.rs [hunk 1]
```diff
@@ -52,7 +52,7 @@ impl<'a> TreeSitterFile<'a> {
             .map_err(|_| TreeSitterFileError::LanguageMismatch)?;
 
         // do not permit files that take >1s to parse
```
Comment (calyptobai) [inline server/bleep/src/intelligence.rs:54]:
```suggestion
        // do not permit files that take >100ms to parse
```

**Comment by nerdypepper** [comment]
>Interesting approach. I'm curious if we can fix the performance issue in an alternate way with the use of a new data structure?

The issue we are currently seeing is fundamental to tree-sitter from what I understand. Executing the tree-sitter-query DSL program (every `scopes.scm` file) simply takes too much time on complex ASTs (`ts_tree_cursor_parent_node` indicates tree-traversal).

File: server/bleep/src/intelligence.rs (hunk 1 above)
Comment (nerdypepper) [inline server/bleep/src/intelligence.rs:54]:
```suggestion
        // do not permit files that take >100ms to parse
```

**Review by nerdypepper — APPROVED** [review]

**Review by nerdypepper — DISMISSED** [review]

**Pull request closed without merging**